core/src/main/java/org/apache/iceberg/avro/SchemaToType.java

```diff
@@ -102,15 +107,47 @@ public Type record(Schema record, List<String> names, List<Type> fieldTypes) {
     return Types.StructType.of(newFields);
   }
 
+  private Boolean isOfType(Type type, Set<Type.TypeID> validTypes) {
+    return type.isPrimitiveType() && validTypes.contains(type.typeId());
+  }
+
   @Override
   public Type union(Schema union, List<Type> options) {
     Preconditions.checkArgument(AvroSchemaUtil.isOptionSchema(union),
         "Unsupported type: non-option union: %s", union);
-    // records, arrays, and maps will check nullability later
-    if (options.get(0) == null) {
-      return options.get(1);
+    if (AvroSchemaUtil.isOptionSchema(union)) {
+      // records, arrays, and maps will check nullability later
+      if (options.get(0) == null) {
+        return options.get(1);
+      } else {
+        return options.get(0);
+      }
     } else {
-      return options.get(0);
+      Set<Type.TypeID> longTypes = new HashSet<>();
+      longTypes.add(Type.TypeID.INTEGER);
+      longTypes.add(Type.TypeID.LONG);
+
+      Set<Type.TypeID> doubleTypes = new HashSet<>();
+      longTypes.add(Type.TypeID.FLOAT);
+      longTypes.add(Type.TypeID.DOUBLE);
+
+      if (options.size() == 1) {
+        return options.get(0);
+      } else if (options.size() == 2 && isOfType(options.get(0), longTypes) && isOfType(options.get(1), longTypes)) {
+        return Types.LongType.get();
+      } else if (options.size() == 2 && isOfType(options.get(0), longTypes) && isOfType(options.get(1), doubleTypes)) {
```

Comment:
The changes for size 1 and Long/Doubles are just trying to line up with the spark-avro SchemaConverter. these cna be removed